exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java

```diff
@@ -0,0 +1,755 @@
+      this.targetOp = targetOp;
+      this.fragment = fragment;
+    }
+
+    @Override
+    public Void visitExchange(Exchange exchange, Void value) throws RuntimeException {
+      List<Fragment.ExchangeFragmentPair> exchangeFragmentPairs = fragment.getReceivingExchangePairs();
+      for (Fragment.ExchangeFragmentPair exchangeFragmentPair : exchangeFragmentPairs) {
+        boolean same = exchange == exchangeFragmentPair.getExchange();
+        if (same) {
+          return null;
+        }
+      }
+      return exchange.getChild().accept(this, value);
+    }
+
+    @Override
+    public Void visitOp(PhysicalOperator op, Void value) throws RuntimeException {
+      boolean same = op == targetOp;
+      if (!same) {
+        for (PhysicalOperator child : op) {
+          child.accept(this, value);
+        }
+      } else {
+        contain = true;
+      }
+      return null;
+    }
+
+    public boolean isContain() {
+      return contain;
+    }
+  }
+
+  private boolean containsPhysicalOperator(Wrapper wrapper, PhysicalOperator op) {
+    WrapperOperatorsVisitor wrapperOpsVistitor = new WrapperOperatorsVisitor(op, wrapper.getNode());
+    wrapper.getNode().getRoot().accept(wrapperOpsVistitor, null);
+    return wrapperOpsVistitor.isContain();
+  }
+
+  private Wrapper findPhysicalOpContainer(Wrapper wrapper, PhysicalOperator op) {
+    boolean contain = containsPhysicalOperator(wrapper, op);
+    if (contain) {
+      return wrapper;
+    }
+    List<Wrapper> dependencies = wrapper.getFragmentDependencies();
+    if (CollectionUtils.isEmpty(dependencies)) {
+      return null;
+    }
+    for (Wrapper dependencyWrapper : dependencies) {
+      Wrapper opContainer = findPhysicalOpContainer(dependencyWrapper, op);
+      if (opContainer != null) {
+        return opContainer;
+      }
+    }
+    //should not be here
+    return null;
+  }
+
+  private static boolean isEqJoin(HashJoinPOP joinPOP) {
+    List<JoinCondition> joinConditions = joinPOP.getConditions();
+    for (JoinCondition joinCondition : joinConditions) {
+      String relationShip = joinCondition.getRelationship();
+      boolean equalRelationShip = "==".equals(relationShip) || "EQUALS".equalsIgnoreCase(relationShip);
+      if (!equalRelationShip) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static class HashJoinOpHolder {
+
+    private int joinOpId;
+
+    private int joinMajorId;
+
+    private HashJoinPOP hashJoinPOP;
+
+    private int probeSideScanMajorId;
+
+    private int probeSideScanOpId;
+
+    private PhysicalOperator probeSideScanOp;
+
+    private List<CoordinationProtos.DrillbitEndpoint> probeSideScanEndpoints;
+
+    //probe side's children have blocked nodes.
+    private boolean meetBlockedOperator;
+
+    //whethe this join operator is a partitioned HashJoin or broadcast HashJoin
+    private boolean meetBroadcastExchange;
+
+    private boolean meetPartitionedExchange;
```

Comment:
Better names would be `encounteredPartitionExchange`,  `encounteredBroadcastExchange`.   But do you really need these variables declared in the HashJoinOpHolder class since these are temporary states encountered during the traversal of the operator tree and you are anyways having another class variable `isBroadcastHashJoin`.